activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb

```diff
@@ -317,6 +317,9 @@ def register_pool(pool, frequency) # :nodoc:
           private
             def spawn_thread(frequency)
               Thread.new(frequency) do |t|
+                # Advise multi-threaded app servers to ignore this thread for
+                # the purposes of fork safety warnings
+                Thread.current.thread_variable_set(:fork_safe, true)
```

Comment:
Instead of setting a thread local variable, I was wondering if we can make use of `Thread#name` instead. That way we can just name this thread with `Thread#name=` and expose the name as a constant which Puma can refer to.

Reply:
From Puma's perspective, doesn't matter to us. I'll do whatever gets it merged.

Reply:
I prefer the thread-local approach, since it allows other frameworks or even applications to silence the warning from Puma if they know the threads they start during boot are fork-safe. Having Puma depend on a constant defined in Rails, when it can run any Rack app, also seems non-ideal.

Reply:
I was thinking Puma would have to check if Rails or ActiveRecord in this case is loaded before excluding the reaper thread during lookup. Whether it is a named thread or local thread variable doesn’t matter imo but I think it is important that Rails remains in control over how to make the thread identifiable to external dependencies and hence my suggestion to define the name in a constant as part of a public API.

Reply:
I don't want to make identifying the reaper thread(s) part of Rails' public API. The way I see it, we're using a (not yet implemented) Puma feature here, not exposing a new API from Rails.

@nateberkopec it might help this along if we could see the implementation in Puma first: when we're happy with both sides, we can turn the keys at the same time, WarGames-style.

![](https://i.imgur.com/F9HLhv2.gif)

Reply:
@eugeneius sure, let me release 5.0.3 on Monday and I'll have something for you after that.